python/src/scipp/compat/dict.py

```diff
@@ -0,0 +1,227 @@
+                for key, attr in scipp_obj.attrs.items():
+                    out[name]["attrs"][key] = _variable_to_dict(attr)
+        return out
+
+
+def _vec_parser(x, shp):
+    """
+    Parse vector_3_float to 2D numpy array
+    """
+    return np.array(x)
+
+
+def _event_parser(x, shp):
+    """
+    Parse event list data to numpy array of numpy arrays
+    """
+    return np.reshape([np.array(x[i]) for i in range(len(x))], shp)
+
+
+def _variable_to_dict(v):
+    """
+    Convert a scipp Variable to a python dict.
+    """
+    out = {
+        "dims": _dims_to_strings(v.dims),
+        "shape": v.shape,
+        "unit": v.unit,
+        "dtype": v.dtype
+    }
+
+    # Use defaultdict to return the raw values/variances by default
+    dtype_parser = defaultdict(lambda: lambda x, y: x)
+    # Using raw dtypes as dict keys doesn't appear to work, so we need to
+    # convert to strings.
+    dtype_parser.update({
+        str(sc.dtype.vector_3_float64): _vec_parser,
+        str(sc.dtype.matrix_3_float64): _vec_parser,
+        str(sc.dtype.string): _vec_parser,
+        str(sc.dtype.event_list_float32): _event_parser,
+        str(sc.dtype.event_list_float64): _event_parser
+    })
+
+    str_dtype = str(v.dtype)
+
+    # Check if variable is 0D:
+    suffix = "s" if len(out["dims"]) > 0 else ""
+    out["value" + suffix] = dtype_parser[str_dtype](getattr(
+        v, "value" + suffix), v.shape)
+    var = getattr(v, "variance" + suffix)
+    out["variance" + suffix] = dtype_parser[str_dtype](
+        var, v.shape) if var is not None else None
+    return out
+
+
+def _data_array_to_dict(da):
+    """
+    Convert a scipp DataArray to a python dict.
+    """
+    out = {"coords": {}, "masks": {}, "attrs": {}}
+    for key in out.keys():
+        for name, item in getattr(da, key).items():
+            out[key][str(name)] = _variable_to_dict(item)
+    if da.unaligned is not None:
+        out["unaligned"] = _data_array_to_dict(da.unaligned)
+    else:
+        out["data"] = _variable_to_dict(da.data)
+    out["name"] = da.name
+    return out
+
+
+def _dims_to_strings(dims):
+    """
+    Convert dims that may or may not be strings to strings.
+    """
+    return [str(dim) for dim in dims]
+
+
+def from_dict(dict_obj):
+    """
+    Convert a python dict to a scipp Variable, DataArray or Dataset.
+    If the input keys contain both `'coords'` and `'data'`, then a DataArray is
+    returned.
+    If the input keys contain both `'dims'` and `'values'`, as Variable is
+    returned.
+    Otherwise, a Dataset is returned.
+
+    :param dict_obj: A python dict to be converted to a scipp object.
+    :type dict_obj: dict
+    :return: A scipp Variable, DataArray or Dataset.
+    :rtype: Variable, DataArray, or Dataset
+    """
+    keys_as_set = set(dict_obj.keys())
+    if ({"coords", "data"}.issubset(keys_as_set)
+            or {"coords", "unaligned"}.issubset(keys_as_set)):
+        # Case of a DataArray-like dict (most-likely)
```

Comment:
xarray appears to define these as class methods, and thus avoids the guessing: http://xarray.pydata.org/en/stable/generated/xarray.DataArray.from_dict.html. Any reasons why or why not do it like that?

Reply:
I did think about making `to_dict` as a method. And then remembered that we discussed we should add as few methods as possible to the classes. So I went for the free function. But it does make more sense to me as a method.
I'm happy to change it.

Reply:
👍